packages/flutter/lib/src/widgets/routes.dart

```diff
@@ -618,6 +636,76 @@ abstract class ModalRoute<T> extends TransitionRoute<T> with LocalHistoryRoute<T
   Animation<double> get forwardAnimation => _forwardAnimationProxy;
   ProxyAnimation _forwardAnimationProxy;
 
+  /// Return the value of the first callback added with
+  /// [addScopedWillPopCallback] that returns false. Otherwise return true.
+  ///
+  /// Typically this method is not overridden because applications usually
+  /// don't create modal routes directly, they use higher level primitives
+  /// like [showDialog]. The scoped [WillPopCallback] list makes it possible
+  /// for ModalRoute descendants to collectively define the value of `willPop`.
+  ///
+  /// See also:
+  ///
+  /// * [Form], which provides an `onWillPop` callback that uses this mechanism.
+  /// * [addScopedWillPopCallback], which adds a callback to the list this
+  ///   method checks.
+  /// * [removeScopedWillPopCallback], which removes a callback from the list
+  ///   this method checks.
+  @override
+  Future<bool> willPop() async {
+    final _ModalScopeState scope = _scopeKey?.currentState;
+    assert(scope != null);
+    for (WillPopCallback callback in new List<WillPopCallback>.from(scope.willPopCallbacks)) {
+      if (!await callback())
+        return false;
+    }
+    return true;
+  }
+
+  /// Enables this route to veto attempts by the user to dismiss it.
+  ///
+  /// This callback is typically added by a stateful descendant of the modal route.
+  /// A stateful widget shown in a modal route, like the child passed to
+  /// [showDialog], can look up its modal route and then add a callback int its
+  /// `dependenciesChanged` method:
+  ///
+  /// ```dart
+  /// @override
+  /// void dependenciesChanged() {
+  ///  super.dependenciesChanged();
+  ///  ModalRoute.of(context).addScopedWillPopCallback(askTheUserIfTheyAreSure);
+  /// }
+  /// ```
+  ///
+  /// A typical application of this callback would be to warn the user about
+  /// unsaved [Form] data if the user attempts to back out of the form.
+  ///
+  /// See also:
+  ///
+  /// * [Form], which provides an `onWillPop` callback that uses this mechanism.
+  /// * [willPop], which runs the callbacks added with this method.
+  /// * [removeScopedWillPopCallback], which removes a callback from the list
+  ///   that [willPop] checks.
+  void addScopedWillPopCallback(WillPopCallback callback) {
```

Comment:
should probably warn that this might get called long after the route and the widget tree is disposed.